### PR TITLE
[FSDP2] Added autograd/memory/overlap/frozen/2D/AC tests

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #119551
* #119378
* #119302
* #118298
* #118755
* #118223
* __->__ #118136
* #119550

This PR adds tests for autograd (mainly backward hooks), memory, overlap, and frozen parameters.
- Autograd: unused forward output, unused forward module, non-tensor activations (common in internal models)
- Memory: expected GPU memory usage after init, forward, backward, and optimizer step
- Overlap: communication/computation overlap in forward and backward
- Frozen: expected reduce-scatter size, training parity

This PR adds some initial 2D (FSDP + TP) training and model state dict tests. The only change required for model sharded state dict is to make sure parameters are sharded before save and load.

This PR adds tests that `fully_shard` can use `torch.utils.checkpoint`, `_composable.checkpoint`, and `CheckpointWrapper` on a transformer.

(I squashed all of these into one PR now to save CI cost.)

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

**Comment by awgu** [comment]
To-do: I think we can beef up unit tests more:
- Add unit test to check unshard/reshard/hook events when composing with AC
- Add unit test to check FSDP-managed memory when composing with AC
- Add unit test to check FSDP-managed memory when using mixed precision

File: test/distributed/_composable/fsdp/test_fully_shard_memory.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        # Use a small input to minimize activation memory usage
+        inp = torch.randint(0, vocab_size, (1, 4), device="cuda")
+
+        # Forward:
+        loss = model(inp)
+        mem_mb = self._get_peak_active_memory_mb()
+        # Allow for some buffer for fragmentation/activations (where this
+        # number is kept much smaller than the actual memory usage, which is on
+        # the order of 100-200+ MB)
+        buffer_mb = 16
```
Comment (weifengpy) [inline test/distributed/_composable/fsdp/test_fully_shard_memory.py:88]: I remember that you mentioned cublas have larger workspace in H100? is it still tolerated within 16MB?

**Review by weifengpy — APPROVED** [review]

File: test/distributed/_composable/fsdp/test_fully_shard_memory.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+            {"reshard_after_forward": [True, False]},
+            self._test_fully_shard_training_memory,
+        )
+
+    def _test_fully_shard_training_memory(self, reshard_after_forward: bool):
+        assert (
+            self.world_size == 2
+        ), f"Requires world size of 2 since some values are hard coded: {self.world_size}"
+        torch.manual_seed(42)
+        # Pre-run a linear forward (gemm and bias) and backward (gemm) to
```
Comment (awgu) [inline test/distributed/_composable/fsdp/test_fully_shard_memory.py:35]: @weifengpy I did this trick here 😄 

I pre-allocate the cuBLAS workspaces and do not count them to the memory by recording the baseline memory after the workspaces have been initialized.

Comment (weifengpy) [reply test/distributed/_composable/fsdp/test_fully_shard_memory.py:35]: nice! the memory can be fully estimated now. amazing!

File: test/distributed/_composable/fsdp/test_fully_shard_memory.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+        # Backward:
+        loss.sum().backward()
+        mem_mb = self._get_peak_active_memory_mb()
+        if reshard_after_forward:
+            # 1x max unsharded block parameters (all-gather), 2.5x max
+            # unsharded block gradients (gradients, reduce-scatter input,
+            # reduce-scatter output), non-block parameters, and other
+            expected_mem_mb = (
+                3.5 * max_unsharded_numel + non_block_numel
```
Comment (awgu) [inline test/distributed/_composable/fsdp/test_fully_shard_memory.py:113]: Changed locally -- will push next time:
```suggestion
                3 * max_unsharded_numel + non_block_numel
```
Since we already include `model_sharded_numel * 4` below as part of the "2x sharded parameters/gradients", we do not need to count it as the 0.5 factor here.

File: test/distributed/_composable/fsdp/test_fully_shard_state_dict.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    ColwiseParallel,
+    parallelize_module,
+    RowwiseParallel,
+)
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_fsdp import FSDPTest, MLP
+from torch.testing._internal.common_utils import run_tests
+
+
+class TestFullyShardStateDict(FSDPTest):
```
Comment (awgu) [inline test/distributed/_composable/fsdp/test_fully_shard_state_dict.py:23]: I only included some basic state dict tests. For example, the 2D state dict tests here do not break even without having strided sharding.

The tests are written mainly to exercise the state dict hooks to force-reshard parameters if saving or loading a state dict.

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 5]
```diff
@@ -149,6 +156,9 @@ def _init_grad_divide_factors(self):
             data_parallel_world_size / self._grad_predivide_factor
         )
 
+    def lazy_init(self):
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:159]: I moved the state dict hook registration to be 'lazy' because I was prototyping the option to fuse FSDP states/parameter groups after construction time and before lazy initialization.

This relies on an assumption that the state dict hooks simply reshard the parameters. (I can add a comment.) In other words, there is no need to have the state dict hooks registered unless we have already lazily initialized and unsharded the parameters once. If a user saves or loads a state dict without lazy initialization, then we do not need the state dict hooks to run, as the parameters are guaranteed to be sharded.

**Review by awgu — COMMENTED** [review]
@wz337 @wanchaol I added some comments regarding to the state dict implementation to help navigate this PR.

File: test/distributed/_composable/fsdp/test_fully_shard_state_dict.py (hunk 4 above)
Comment (wz337) [reply test/distributed/_composable/fsdp/test_fully_shard_state_dict.py:23]: > I only included some basic state dict tests. For example, the 2D state dict tests here do not break even without having strided sharding.
> 
> The tests are written mainly to exercise the state dict hooks to force-reshard parameters if saving or loading a state dict.

2D state_dict should be working right now because you are sharding the tensors from the parent_mesh directly. We can do this because we know that FSDP is the outermost dimension and the state_dict call should come from the outermost parallelism. 

Ideally, the flow would be a given parallelism should not need to inquire the parent_mesh to produce the globally sound DTensor.  This would be more composable friendly, as then in theory the order of parallelism would not matter. To achieve this flexibility, I believe we will need the `distribute_tensor` API to accept DTensor and that may require the strided sharding that @wanchaol mentioned.

File: test/distributed/_composable/fsdp/test_fully_shard_frozen.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    FSDPTest,
+    MLP,
+    patch_reduce_scatter,
+    patch_register_post_backward_hook_backward,
+    reduce_scatter_with_assert,
+)
+from torch.testing._internal.common_utils import run_tests
+
+
+class TestFullyShardFrozen(FSDPTest):
```
Comment (awgu) [inline test/distributed/_composable/fsdp/test_fully_shard_frozen.py:30]: Candidate for multi-GPU tests

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 7]
```diff
@@ -377,6 +389,94 @@ def forward(self, x):
             self.assertEqual(losses[0], losses[1])
 
 
+class TestFullyShard1DTrainingCompose(FSDPTest):
+    @property
+    def world_size(self) -> int:
+        # Since these tests run with a larger transformer model, they may see
+        # some numeric drift with >2 GPUs
+        return 2
```
Comment (awgu) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py:397]:
```suggestion
        return min(torch.cuda.device_count(), 2)
```

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 8]
```diff
@@ -377,6 +389,94 @@ def forward(self, x):
...
+
+    @skip_if_lt_x_gpu(2)
+    def test_train_parity_with_activation_checkpointing(self):
```
Comment (awgu) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py:400]: Candidate for multi-GPU tests

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 9]
```diff
@@ -423,5 +523,81 @@ def _test_train_shared_params(
             self.assertEqual(losses[0], losses[1])
 
 
+class TestFullyShard2DTraining(FSDPTest):
+    @property
+    def world_size(self) -> int:
+        return min(4, torch.cuda.device_count())
+
+    @skip_if_lt_x_gpu(2)
+    def test_train_parity_2d_mlp(self):
```
Comment (awgu) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py:532]: Candidate for multi-GPU tests

File: test/distributed/_composable/fsdp/test_fully_shard_state_dict.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+                self.assertEqual(
+                    tensor._local_tensor.data_ptr(), param_name_to_data_ptr[param_name]
+                )
+
+        # Verify that we can load a new state dict that contains DTensors with
+        # storages different from the current model parameters
+        new_state_dict: Dict[str, DTensor] = {}
+        for param_name, dtensor in state_dict.items():
+            # Construct new DTensors to exercise load state dict writeback
+            new_state_dict[param_name] = dtensor.detach().clone().fill_(new_fill_value)
```
Comment (wanchaol) [inline test/distributed/_composable/fsdp/test_fully_shard_state_dict.py:139]: I think this is essentially testing the same sharded parameter can be saved/loaded in the same way? i.e. load state_dict essentially just load back a manually written value. 

This is working because **both** 2D state_dict and the new model state_dict generates the same strided sharding, so directly loading with the same topology works.

IMO for 2d checkpoint load re-sharding, it does not work, suppose we have a module that have only one parameter:

```
param = torch.arange(4) + 5
```
after TP (assuming TP Shard(0)), it become:
```
rank 0, 2: [5, 6]
rank 1, 3: [7, 8]
```
Then after FSDP it would become:
```
rank 0: [5]
rank 1: [7]
rank 2: [6]
rank 3: [8]
```
DCP would write offsets as if it would be:
```
rank 0: [5]
rank 1: [6]
rank 2: [7]
rank 3: [8]
```
Then resharding would incorrectly load the data to the wrong offset

cc @wz337

**Review by wanchaol — APPROVED** [review]
Tests looks good to me! I think 2D state_dict tests works but it might not work for resharding case, where we need to add strided sharding support.

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 11]
```diff
@@ -423,5 +523,81 @@ def _test_train_shared_params(
...
+            "cuda", (dp_size, self.world_size // dp_size), mesh_dim_names=("dp", "tp")
+        )
+        self.run_subtests(
+            {
+                "reshard_after_forward": [False, True],
+                "use_activation_checkpointing": [False, True],
+                "mlp_dim": [3, 16, 17],
+            },
+            functools.partial(self._test_train_parity_2d_mlp, global_mesh),
+        )
```
[the remark this answers is not in the document]
Comment (awgu) [reply test/distributed/_composable/fsdp/test_fully_shard_training.py:545]: The main reason I have been subtesting by defining a separate `_test...` method is so that we do not have to indent the entire test code due to:
```
for arg1, arg2, ... in itertools.product(...):
    # Actual test code
```
The `run_subtests` helper that hides that `itertools.product` (or similar) takes in a method, which is why there is a separate `_test_train_parity_2d_mlp` method 😢 .

If you have a better solution for how to do subtesting elegantly, I would be happy to migrate the FSDP2 test suite.

[the remark this answers is not in the document]
Comment (awgu) [reply test/distributed/_composable/fsdp/test_fully_shard_training.py:545]: Also, I have the `global_mesh = init_device_mesh()` only once to reuse across all subtests to avoid the expensive `init_process_group()` calls upon each subtest. Once we re-land the process group caching, this would be able to be simplified (pushing the `global_mesh = init_device_mesh()` call into `_test_train_parity_2d_mlp`).
